### DISCUSSION TITLE
Orientation with the use of Invoke

### DISCUSSION THREAD

**Discussion by manoelpqueiroz** [getpelican/pelican#3365] [Q&A] [answered]
Hi guys, I have a question not related to the project per se, but curious to get some best practices with the `invoke` library you use as a dependency.

I really liked the fact you have set several constants in the [`tasks.py`](https://github.com/getpelican/pelican/blob/main/tasks.py) and I would like to better understand the reasoning behind this choice.

I have basically three questions:

- How does implementing those constants actually contribute to making development easier?
- What is the advantage of calling Python modules with `{VENV_BIN}/module`? Why use `{VENV_BIN}/pytest` instead of just the `pytest` callable in [line 51](https://github.com/getpelican/pelican/blob/513abbfdc668946590194c637dbe90ec228aaf6f/tasks.py#L51), for instance?
- What is the purpose of the `PTY` constant in [line 11](https://github.com/getpelican/pelican/blob/513abbfdc668946590194c637dbe90ec228aaf6f/tasks.py#L11) to check whether the OS is Windows or not and how does it influence when applied in every command, like [here](https://github.com/getpelican/pelican/blob/513abbfdc668946590194c637dbe90ec228aaf6f/tasks.py#L44) or [here](https://github.com/getpelican/pelican/blob/513abbfdc668946590194c637dbe90ec228aaf6f/tasks.py#L94)?

While not directly related to the project itself, I really appreciate the help, it serves as a great opportunity for me to learn about best practices in Python development.

**Comment by justinmayer** [comment, marked as answer]
Hi Manoel. Welcome! I appreciate the very thoughtful tone in which you posed your questions. In no particular order:

- Invoke is by no means a dependency, but rather a convenience. It is included among the _development_ dependencies to make it easier for folks to get up-and-running via `invoke setup`, `invoke tests`, `invoke lint`, etc.
- Some of the constants are an attempt to keep code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), such as the `PTY` one you mentioned, while others are probably a form of premature optimization in that they may or may not be sufficiently utilized in subsequent tasks.
- If you look at how `VENV_BIN` is composed, you may notice two implicit purposes: (1) Invoke the specified executable from within the _active_ virtual environment if one is indeed active; otherwise use the default path of `~/virtualenvs/pelican/[bin|Scripts]/[executable_name]`. (2) As implied in the preceding sentence, use `/bin/` on MacOS/Linux and `/Scripts/` on Windows.
- Why not just run `pytest`, `ruff`, etc. and let `$PATH` sort it out? Because sometimes the versions of those tools need to be controlled to ensure tests and linters function properly with the current state of Pelican's code, so we can't just rely on whatever version a user has globally available on `$PATH` and must instead insist that the project-specified version be called from within a Pelican-specific virtual environment.
- The purpose of the `PTY` constant is to ensure that output looks similar to what it would be if you were to run the same command by itself, outside of an `invoke […]` context. For example, without `PTY = True`, colors and other rich formatting would otherwise be absent in the output when calling many of the specified executables from within Invoke tasks.
- I don't use Windows, but `PTY` should be `False` for proper behavior on Windows, if I understand what @MinchinWeb has said about it in the past.

I hope this helps!

**Comment by manoelpqueiroz** [comment]
Hi Justin, likewise, thank you very much for your explanations! It's a small piece of the project and although it does not have direct impact on the codebase, I really appreciate the simplicity and elegance of the solution.

I'm studying Invoke to implement it in the [cookiecutter template](https://www.gitlab.com/manoelpqueiroz/galactipy) I'm working on and I will be definitely using your solution as a best practice.

Thank you once again!
